### ISSUE TITLE
Unnecessarily scaling gradients when gradient_accumulation_steps is 1

### ISSUE THREAD

**Issue by shunting314** [pytorch/torchtune#2515]
The `scale_grads` call [here](https://github.com/pytorch/torchtune/blob/124123149c0fedc59c889da64c385df0f1a4d830/recipes/full_finetune_single_device.py#L706) will need to read and write each parameter once. For a 8B parameters model with bfloat16 dtype, that results in 8B * 2byte * 2 = 32GB memory access. Assuming a 2TBGS memory bandwidth, this will translate to 16ms latency. Note that this estimation is a lower bound. In practice, due to kernel launch overhead, this latency can be as larger as 42.5ms:

<img width="1169" alt="Image" src="https://github.com/user-attachments/assets/f160a3b4-3540-46f0-8c5e-5859d7f9a31f" /> (check the long blue bar)

Simple change like
```
diff --git a/recipes/full_finetune_single_device.py b/recipes/full_finetune_single_device.py
index 39e596ae..83007343 100644
--- a/recipes/full_finetune_single_device.py
+++ b/recipes/full_finetune_single_device.py
@@ -692,14 +692,21 @@ class FullFinetuneRecipeSingleDevice(FTRecipeInterface):

                 # Loss is normalized by default so we multiply by the number of tokens
                 # This way we can normalize by the total number of tokens if we're accumulating gradients
-                current_loss = self._loss_step(batch) * current_num_tokens
-                running_loss += current_loss
-                current_loss.backward()
+
+                if self._gradient_accumulation_steps > 1:
+                    current_loss = self._loss_step(batch) * current_num_tokens
+                    running_loss += current_loss
+                    current_loss.backward()
+                else:
+                    current_loss = self._loss_step(batch)
+                    running_loss += current_loss * current_num_tokens
+                    current_loss.backward()

                 # Step with optimizer
                 if (idx + 1) % self._gradient_accumulation_steps == 0:
                     if not self._optimizer_in_bwd:
-                        training.scale_grads(self._model, 1 / num_tokens)
+                        if self._gradient_accumulation_steps > 1:
+                            training.scale_grads(self._model, 1 / num_tokens)
                         if self._clip_grad_norm is not None:
                             grad_norm = torch.nn.utils.clip_grad_norm_(
                                 self._model.parameters(),
```

can just skip this `scale_grads` call when gradient accumulation steps is 1.

**Comment by nathan-az** [comment]
Just a note if there's appetite for this enhancement - in the gradient accumulation case, if we know the token count for each update step ahead of time, I think the loss can be scaled correctly before each backward step, completely avoid the separate scaling of the grad norms.

I don't know that this would be straightforward without a pretty significant refactor of the training loop. Below is an example.

<details>

```py
for update_step in range(self._steps_per_epoch):
    # pre-fetch examples and move to device
    batches = []
    token_counts = []
    for _ in range(self._gradient_accumulation_steps):
        batch = next(self._dataloader)
        utils.batch_to_device(batch, self._device)
        current_num_tokens = (
            batch["labels"] != self._loss_fn.ignore_index
        ).sum()
        num_tokens += current_num_tokens
        batches.append(batch)
        token_counts.append(current_num_tokens.item())
        
    torch.distributed.all_reduce(num_tokens)

    for token_count, batch in zip(token_counts, batches):
        labels = batch.pop("labels")
        # prepare inputs
        current_loss = self._loss_fn(logits, labels) * (self.dp_size * token_count / num_tokens)
        del logits
        running_loss += current_loss

    current_loss.backward()
    
    if not self._optimizer_in_bwd:
        self._optimizer.step()
        self._optimizer.zero_grad(set_to_none=True)
```

</details>

I'm not certain the above is correct (and it's not robust, would need handling for things like `StopIteration` since we don't directly iterate through the dataloader). However I think it's correct and removes the scaling. In addition, there might be some slight precision benefit to scaling the loss this way, since `token_count / num_tokens` should be less extreme than accumulating the sum?
